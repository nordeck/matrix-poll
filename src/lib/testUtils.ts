/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  PowerLevelsStateEvent,
  RoomEvent,
  RoomMemberStateEventContent,
  StateEvent,
} from '@matrix-widget-toolkit/api';
import {
  GroupContent,
  IPoll,
  IPollSettings,
  IVote,
  PollStartEvent,
  PollType,
  ResultType,
  RoomNameEvent,
} from '../model';

/**
 * Create a matrix room member event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockRoomMember({
  state_key = '@user-alice',
  event_id = '$event-id-0',
  content = {},
}: {
  state_key?: string;
  event_id?: string;
  content?: Partial<RoomMemberStateEventContent>;
} = {}): StateEvent<RoomMemberStateEventContent> {
  return {
    type: 'm.room.member',
    sender: '@user-id',
    content: {
      membership: 'join',
      displayname: 'Alice',
      avatar_url: 'mxc://alice.png',
      ...content,
    },
    state_key,
    origin_server_ts: 0,
    event_id,
    room_id: '!room-id',
  };
}

/**
 * Create a matrix room name event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockRoomName({
  room_id = '!room-id',
  content = {},
}: {
  room_id?: string;
  content?: Partial<RoomNameEvent>;
} = {}): StateEvent<RoomNameEvent> {
  return {
    type: 'm.room.name',
    sender: '',
    content: {
      name: 'My Room',
      ...content,
    },
    state_key: room_id,
    origin_server_ts: 0,
    event_id: '$event-id-0',
    room_id,
  };
}

/**
 * Create a matrix power levels event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockPowerLevelsEvent({
  content = {},
}: {
  content?: Partial<PowerLevelsStateEvent>;
} = {}): StateEvent<PowerLevelsStateEvent> {
  return {
    type: 'm.room.power_levels',
    sender: '@user-id',
    content: {
      users_default: 100,
      ...content,
    },
    state_key: '',
    origin_server_ts: 0,
    event_id: '$event-id-0',
    room_id: '!room-id',
  };
}

/**
 * Create a poll event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockPoll({
  state_key = 'poll-0',
  event_id = '$event-id-0',
  origin_server_ts = 0,
  content = {},
}: {
  state_key?: string;
  event_id?: string;
  origin_server_ts?: number;
  content?: Partial<IPoll>;
} = {}): StateEvent<IPoll> {
  return {
    type: 'net.nordeck.poll',
    sender: '@user-id',
    content: {
      title: 'My Title',
      question: 'My Question',
      description: 'My Description',
      pollType: PollType.Open,
      answers: [
        { id: '1', label: 'Yes' },
        { id: '2', label: 'No' },
      ],
      resultType: ResultType.Visible,
      duration: 1,
      groups: [
        {
          id: 'red-party',
          eventId: '$event-id-0',
          abbreviation: 'Red Party',
          color: '#ff0000',
          votingRights: {
            '@user-alice': {
              state: 'active',
            },
            '@user-bob': {
              state: 'active',
            },
            '@user-charlie': {
              state: 'active',
            },
          },
        },
        {
          id: 'blue-party',
          eventId: '$event-id-1',
          abbreviation: 'Blue Party',
          color: '#0000ff',
          votingRights: {
            '@user-dameon': {
              state: 'active',
            },
          },
        },
      ],
      ...content,
    },
    state_key,
    origin_server_ts,
    event_id,
    room_id: '!room-id',
  };
}

/**
 * Create a poll start event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockPollStart({
  event_id = '$start-event-id',
  origin_server_ts = 0,
  content = {},
}: {
  state_key?: string;
  event_id?: string;
  origin_server_ts?: number;
  content?: Partial<PollStartEvent>;
} = {}): RoomEvent<PollStartEvent> {
  return {
    type: 'net.nordeck.poll.start',
    sender: '@user-id',
    content: {
      ...content,
    },
    origin_server_ts,
    event_id,
    room_id: '!room-id',
  };
}

/**
 * Create a group event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockGroup({
  state_key = 'group-0',
  event_id = '$event-id-0',
  content = {},
}: {
  state_key?: string;
  event_id?: string;
  content?: Partial<GroupContent>;
} = {}): StateEvent<GroupContent> {
  return {
    type: 'net.nordeck.poll.group',
    sender: '@user-id',
    content: {
      abbreviation: 'GROUP 0',
      color: '#07f556',
      members: {},
      ...content,
    },
    state_key,
    origin_server_ts: 0,
    event_id,
    room_id: '!room-id',
  };
}

/**
 * Create a matrix vote event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockVote({
  event_id,
  sender = '@user-bob',
  origin_server_ts = 0,
  content = {},
}: {
  event_id?: string;
  sender?: string;
  origin_server_ts?: number;
  content?: Partial<IVote>;
} = {}): RoomEvent<IVote> {
  return {
    type: 'net.nordeck.poll.vote',
    sender,
    content: {
      answerId: '1',
      pollId: 'poll-0',
      'm.relates_to': {
        rel_type: 'm.reference',
        event_id: '$start-event-id',
      },
      ...content,
    },
    origin_server_ts,
    event_id: event_id ?? sender,
    room_id: '!room-id',
  };
}

/**
 * Create a poll settings event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockPollSettings({
  event_id = '$event-id-0',
  content = {},
}: {
  event_id?: string;
  content?: Partial<IPollSettings>;
} = {}): StateEvent<IPollSettings> {
  return {
    type: 'net.nordeck.poll.settings',
    sender: '@user-id',
    content: {
      ...content,
    },
    state_key: '!room-id',
    origin_server_ts: 0,
    event_id,
    room_id: '!room-id',
  };
}
