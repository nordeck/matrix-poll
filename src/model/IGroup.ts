/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { StateEvent } from '@matrix-widget-toolkit/api';
import Joi from 'joi';
import { omit } from 'lodash';
import { isValidEvent } from './validation';

export const STATE_EVENT_POLL_GROUP = 'net.nordeck.poll.group';

export type MemberRole = 'delegate' | 'representative';

export interface MemberContent {
  memberRole: MemberRole;
}

export interface GroupContent {
  abbreviation: string;
  color: string;
  members: { [key: string]: MemberContent };
}

export type IGroup = StateEvent<GroupContent>;

const memberContentSchema = Joi.object<MemberContent, true>({
  // Optional for backward compatibility, is resolved during migration
  memberRole: Joi.string().valid('delegate', 'representative').optional(),
}).unknown();

const groupContentSchema = Joi.object<GroupContent, true>({
  abbreviation: Joi.string().min(1).required(),
  color: Joi.string().min(1).required(),
  members: Joi.object().pattern(Joi.string(), memberContentSchema).required(),
}).unknown();

export function isValidGroupEvent(
  event: StateEvent<unknown>,
): event is StateEvent<GroupContent> {
  return isValidEvent(event, STATE_EVENT_POLL_GROUP, groupContentSchema);
}

/**
 * Migrate from the old format of poll groups to the new one.
 *
 * Should be done on every read from the room.
 */
export function migratePollGroupSchema(
  pollGroupEvent: StateEvent<GroupContent>,
): StateEvent<GroupContent> {
  return {
    ...pollGroupEvent,
    content: {
      ...pollGroupEvent.content,
      members: Object.fromEntries(
        Object.entries(pollGroupEvent.content.members)
          // skip all member that have the obsolete `leaveDate` field set.
          .filter(([_, p]) => {
            if ('leaveDate' in p) {
              const { leaveDate } = p as unknown as {
                leaveDate: string;
              };

              return leaveDate === undefined;
            }

            return true;
          })
          .map(([userId, p]) => [
            userId,
            {
              // delete the obsolete fields
              ...omit(p, 'joinDate', 'leaveDate'),
              // make sure we can handle old groups that don't include a memberRole yet
              memberRole: p.memberRole ?? 'delegate',
            },
          ]),
      ),
    },
  };
}
