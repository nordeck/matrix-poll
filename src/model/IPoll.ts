/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { StateEvent } from '@matrix-widget-toolkit/api';
import Joi from 'joi';
import { DateTime } from 'luxon';
import { isValidEvent } from './validation';

export const STATE_EVENT_POLL = 'net.nordeck.poll';

export enum PollType {
  Secret = 'secret',
  Open = 'open',
  ByName = 'byName',
}

export enum ResultType {
  Visible = 'visible',
  Invisible = 'invisible',
}

export interface IPollAnswer {
  id: string;
  label: string;
}

const pollAnswerSchema = Joi.object<IPollAnswer>({
  id: Joi.string().min(1).required(),
  label: Joi.string().min(1).required(),
}).unknown();

export type VotingRight =
  | {
      state: 'active';
    }
  | {
      state: 'invalid';
    }
  | {
      state: 'represented';
      representedBy: string;
    };

const votingRightBase = Joi.object<VotingRight, true>({
  state: Joi.string().valid('active', 'invalid').required(),
}).unknown();

const votingRightRepresented = Joi.object<VotingRight, true>({
  state: Joi.string().valid('represented').required(),
  representedBy: Joi.string().min(1).required(),
}).unknown();

const votingRightSchema = Joi.alternatives().conditional('.state', {
  switch: [
    { is: 'active', then: votingRightBase },
    { is: 'invalid', then: votingRightBase },
    { is: 'represented', then: votingRightRepresented },
  ],
});

export type VotingRights = Partial<Record<string, VotingRight>>;

export type PollGroup = {
  id: string;
  eventId: string;
  abbreviation: string;
  color: string;
  votingRights: VotingRights;
};

const pollGroupSchema = Joi.object<PollGroup, true>({
  id: Joi.string().min(1).required(),
  eventId: Joi.string().min(1).required(),
  abbreviation: Joi.string().min(1).required(),
  color: Joi.string().min(1).required(),
  votingRights: Joi.object()
    .pattern(Joi.string(), votingRightSchema)
    .required(),
}).unknown();

export interface IPoll {
  title: string;
  question: string;
  description: string;
  pollType: PollType;
  answers: Array<IPollAnswer>;
  duration: number;
  resultType: ResultType;
  startEventId?: string;
  startTime?: string;
  endTime?: string;
  groups?: PollGroup[];
}

const pollSchema = Joi.object<IPoll, true>({
  title: Joi.string().min(1).required(),
  question: Joi.string().min(1).required(),
  description: Joi.string().min(0).required(),
  pollType: Joi.string()
    .valid(...Object.values(PollType))
    .required(),
  answers: Joi.array().items(pollAnswerSchema).min(1).required(),
  duration: Joi.number().strict().required(),
  resultType: Joi.string()
    .valid(...Object.values(ResultType))
    .required(),
  startEventId: Joi.string(),
  startTime: Joi.string().isoDate(),
  endTime: Joi.string().isoDate(),
  groups: Joi.array().items(pollGroupSchema),
}).unknown();

export function isValidPollEvent(
  event: StateEvent<unknown>,
): event is StateEvent<IPoll> {
  return isValidEvent(event, STATE_EVENT_POLL, pollSchema);
}

/**
 * Migrate from the old format of polls to the new one.
 *
 * Should be done on every read from the room.
 */
export function migratePollSchema(
  pollEvent: StateEvent<IPoll>,
): StateEvent<IPoll> {
  if (
    pollEvent.content.startTime !== undefined &&
    pollEvent.content.endTime === undefined
  ) {
    return {
      ...pollEvent,
      content: {
        ...pollEvent.content,
        endTime: DateTime.fromISO(pollEvent.content.startTime)
          .plus({ minute: pollEvent.content.duration })
          .toISO(),
      },
    };
  }

  return pollEvent;
}
