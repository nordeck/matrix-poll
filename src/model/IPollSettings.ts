/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { StateEvent } from '@matrix-widget-toolkit/api';
import Joi from 'joi';
import { isValidEvent } from './validation';

export const STATE_EVENT_POLL_SETTINGS = 'net.nordeck.poll.settings';

export interface IPollSettings {
  pdfButtonDisabledAfter?: string;
  pollsOrder?: Array<string>;
}

const settingsSchema = Joi.object<IPollSettings, true>({
  pdfButtonDisabledAfter: Joi.string().isoDate(),
  pollsOrder: Joi.array().items(Joi.string()),
}).unknown();

export function isValidPollSettingsEvent(
  event: StateEvent<unknown>,
): event is StateEvent<IPollSettings> {
  return isValidEvent(event, STATE_EVENT_POLL_SETTINGS, settingsSchema);
}
