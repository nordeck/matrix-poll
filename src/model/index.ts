/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export {
  STATE_EVENT_POLL_GROUP,
  isValidGroupEvent,
  migratePollGroupSchema,
} from './IGroup';
export type { GroupContent, IGroup, MemberContent, MemberRole } from './IGroup';
export {
  PollType,
  ResultType,
  STATE_EVENT_POLL,
  isValidPollEvent,
  migratePollSchema,
} from './IPoll';
export type { IPoll, IPollAnswer, PollGroup, VotingRight } from './IPoll';
export {
  STATE_EVENT_POLL_SETTINGS,
  isValidPollSettingsEvent,
} from './IPollSettings';
export type { IPollSettings } from './IPollSettings';
export { ROOM_EVENT_VOTE, isValidVoteEvent } from './IVote';
export type { IVote } from './IVote';
export { ROOM_EVENT_POLL_START } from './pollStartEvent';
export type { PollStartEvent } from './pollStartEvent';
export { STATE_EVENT_ROOM_NAME, isValidRoomNameEvent } from './roomNameEvent';
export type { RoomNameEvent } from './roomNameEvent';
