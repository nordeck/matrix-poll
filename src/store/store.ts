/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { WidgetApi } from '@matrix-widget-toolkit/api';
import { configureStore, Middleware } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { baseApi } from './api';

const loggerMiddleware: Middleware = () => (next) => (action: unknown) => {
  if (process.env.NODE_ENV === 'development') {
    console.log('LOGGER', action);
  }
  return next(action);
};

export function createStore({
  widgetApi,
  middlewares = [],
}: {
  widgetApi: WidgetApi;
  middlewares?: Middleware[];
}) {
  const store = configureStore({
    reducer: {
      [baseApi.reducerPath]: baseApi.reducer,
    },
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        thunk: {
          extraArgument: {
            widgetApi,
          } as ThunkExtraArgument,
        },
      })
        .concat(middlewares)
        .concat(loggerMiddleware)
        .concat(baseApi.middleware),
  });
  return store;
}

/**
 * Extra arguments that are provided to `createAsyncThunk`
 */
export type ThunkExtraArgument = {
  widgetApi: WidgetApi;
};

type StoreType = ReturnType<typeof createStore>;

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<StoreType['getState']>;
export type AppDispatch = StoreType['dispatch'];

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
