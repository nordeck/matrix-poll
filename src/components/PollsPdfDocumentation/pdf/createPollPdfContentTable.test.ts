/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { mockPoll } from '../../../lib/testUtils';
import { PollType } from '../../../model';
import { PollInvalidAnswer, SelectPollResults } from '../../../store';
import {
  addBreakLineToTheName,
  createPollPdfContentTable,
} from './createPollPdfContentTable';

function mockResults(pollType: PollType): SelectPollResults {
  return {
    poll: mockPoll({
      content: {
        pollType,
        startTime: '2020-01-01T03:33:55Z',
        endTime: '2020-01-01T03:34:55Z',
      },
    }),
    results: {
      votes: {
        '@user-1': '1',
        '@user-2': PollInvalidAnswer,
        '@user-3': '2',
      },
    },
    votingRights: ['@user-1', '@user-2'],
  };
}

function mockGroupedResults(pollType: PollType): SelectPollResults {
  return {
    poll: mockPoll({
      content: {
        pollType,
        startTime: '2020-01-01T03:33:55Z',
      },
    }),
    results: {
      votes: {
        '@user-1': '1',
        '@user-2': PollInvalidAnswer,
        '@user-3': '2',
      },
    },
    votingRights: ['@user-1', '@user-2'],
    groupedResults: {
      'group-1': {
        abbreviation: 'Group 1',
        color: '#ff0000',
        votes: {
          '@user-1': '1',
          '@user-2': PollInvalidAnswer,
        },
        invalidVoters: {},
      },
      'group-2': {
        abbreviation: 'Group 2',
        color: '#0000ff',
        votes: {
          '@user-3': '2',
        },
        invalidVoters: {},
      },
      'group-3': {
        abbreviation: 'Group 3',
        color: '#ffaaff',
        votes: {},
        invalidVoters: {},
      },
    },
  };
}

describe('createPollPdfContentTable', () => {
  const getUserDisplayName = (id: string) => `Name of ${id}`;

  it('should generate table for "byName" with groups', () => {
    const pollResult = mockGroupedResults(PollType.ByName);

    expect(
      createPollPdfContentTable(pollResult, getUserDisplayName),
    ).toMatchSnapshot();
  });

  it('should generate table for "byName" without groups', () => {
    const pollResult = mockResults(PollType.ByName);

    expect(
      createPollPdfContentTable(pollResult, getUserDisplayName),
    ).toMatchSnapshot();
  });

  it('should generate table for "open" with groups', () => {
    const pollResult = mockGroupedResults(PollType.Open);

    expect(
      createPollPdfContentTable(pollResult, getUserDisplayName),
    ).toMatchSnapshot();
  });

  it('should not generate table for "open" without groups', () => {
    const pollResult = mockResults(PollType.Open);

    expect(createPollPdfContentTable(pollResult, getUserDisplayName)).toEqual(
      [],
    );
  });

  it('should not generate table for "secret" with groups', () => {
    const pollResult = mockGroupedResults(PollType.Secret);

    expect(createPollPdfContentTable(pollResult, getUserDisplayName)).toEqual(
      [],
    );
  });

  it('should not generate table for "secret" without groups', () => {
    const pollResult = mockResults(PollType.Secret);

    expect(createPollPdfContentTable(pollResult, getUserDisplayName)).toEqual(
      [],
    );
  });

  it('should add to the name a break line every 45 character if there is 3 answer columns', () => {
    let userName: string;

    // no break line
    userName = 'user-1(@someid.that.break.one.time)';
    expect(addBreakLineToTheName(userName, 3)).toBe(
      `user-1(@someid.that.break.one.time)`,
    );

    // one break line
    userName = 'user-1(@someid.that.break.one.time.that.break.one.time)';
    expect(addBreakLineToTheName(userName, 3)).toBe(
      'user-1(@someid.that.break.one.time.that.break\n.one.time)',
    );

    // two break line
    userName =
      'user-1(@someid.that.break.one.time.that.break.one.time.that.break.one.time.that.break.one.time)';
    expect(addBreakLineToTheName(userName, 3)).toBe(
      'user-1(@someid.that.break.one.time.that.break\n.one.time.that.break.one.time.that.break.one.\ntime)',
    );
  });

  it('should add to the name a break line every 40 character if there is 4 answer columns', () => {
    let userName: string;

    // no break line
    userName = 'user-1(@someid.that.break.one.time)';
    expect(addBreakLineToTheName(userName, 4)).toBe(
      `user-1(@someid.that.break.one.time)`,
    );

    // one break line
    userName = 'user-1(@someid.that.break.one.time.that.break.one.time)';
    expect(addBreakLineToTheName(userName, 4)).toBe(
      'user-1(@someid.that.break.one.time.that.\nbreak.one.time)',
    );

    // two break line
    userName =
      'user-1(@someid.that.break.one.time.that.break.one.time.that.break.one.time.that.break.one.time)';
    expect(addBreakLineToTheName(userName, 4)).toBe(
      'user-1(@someid.that.break.one.time.that.\nbreak.one.time.that.break.one.time.that.\nbreak.one.time)',
    );
  });
});
