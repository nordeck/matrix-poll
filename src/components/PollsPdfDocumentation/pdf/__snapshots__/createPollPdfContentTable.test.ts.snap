// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createPollPdfContentTable > should generate table for "byName" with groups 1`] = `
[
  [
    {
      "table": {
        "body": [
          [
            {
              "columns": [
                {
                  "canvas": [
                    {
                      "color": "#ff0000",
                      "r1": 15,
                      "r2": 15,
                      "type": "ellipse",
                      "x": 15,
                      "y": 15,
                    },
                  ],
                  "width": 33,
                },
                {
                  "relativePosition": {
                    "y": 7.777777777777778,
                  },
                  "style": {
                    "alignment": "left",
                    "bold": true,
                    "fontSize": 13,
                  },
                  "text": "Group 1",
                },
              ],
            },
            {
              "style": "tableHeader",
              "text": "Yes
1",
              "type": "none",
            },
            {
              "style": "tableHeader",
              "text": "No
0",
              "type": "none",
            },
            {
              "style": "tableHeader",
              "text": "Invalid
1",
              "type": "none",
            },
          ],
          [
            {
              "style": {
                "alignment": "left",
              },
              "text": "Name of @user-1",
            },
            {
              "style": "tableBody",
              "text": "X",
            },
            {
              "style": "tableBody",
              "text": "",
            },
            {
              "style": "tableBody",
              "text": "",
            },
          ],
          [
            {
              "style": {
                "alignment": "left",
              },
              "text": "Name of @user-2",
            },
            {
              "style": "tableBody",
              "text": "",
            },
            {
              "style": "tableBody",
              "text": "",
            },
            {
              "style": "tableBody",
              "text": "X",
            },
          ],
        ],
        "headerRows": 1,
        "heights": 18,
        "widths": [
          "66%",
          "10%",
          "10%",
          "14%",
        ],
      },
    },
    {
      "table": {
        "body": [
          [
            {
              "columns": [
                {
                  "canvas": [
                    {
                      "color": "#0000ff",
                      "r1": 15,
                      "r2": 15,
                      "type": "ellipse",
                      "x": 15,
                      "y": 15,
                    },
                  ],
                  "width": 33,
                },
                {
                  "relativePosition": {
                    "y": 7.777777777777778,
                  },
                  "style": {
                    "alignment": "left",
                    "bold": true,
                    "fontSize": 13,
                  },
                  "text": "Group 2",
                },
              ],
            },
            {
              "style": "tableHeader",
              "text": "Yes
0",
              "type": "none",
            },
            {
              "style": "tableHeader",
              "text": "No
1",
              "type": "none",
            },
            {
              "style": "tableHeader",
              "text": "Invalid
0",
              "type": "none",
            },
          ],
          [
            {
              "style": {
                "alignment": "left",
              },
              "text": "Name of @user-3",
            },
            {
              "style": "tableBody",
              "text": "",
            },
            {
              "style": "tableBody",
              "text": "X",
            },
            {
              "style": "tableBody",
              "text": "",
            },
          ],
        ],
        "headerRows": 1,
        "heights": 18,
        "widths": [
          "66%",
          "10%",
          "10%",
          "14%",
        ],
      },
    },
    {
      "table": {
        "body": [
          [
            {
              "columns": [
                {
                  "canvas": [
                    {
                      "color": "#ffaaff",
                      "r1": 15,
                      "r2": 15,
                      "type": "ellipse",
                      "x": 15,
                      "y": 15,
                    },
                  ],
                  "width": 33,
                },
                {
                  "relativePosition": {
                    "y": 7.777777777777778,
                  },
                  "style": {
                    "alignment": "left",
                    "bold": true,
                    "fontSize": 13,
                  },
                  "text": "Group 3",
                },
              ],
            },
            {
              "style": "tableHeader",
              "text": "Yes
0",
              "type": "none",
            },
            {
              "style": "tableHeader",
              "text": "No
0",
              "type": "none",
            },
            {
              "style": "tableHeader",
              "text": "Invalid
0",
              "type": "none",
            },
          ],
          [
            {
              "colSpan": 4,
              "style": {
                "alignment": "center",
              },
              "text": "No votes were cast.",
            },
          ],
        ],
        "headerRows": 1,
        "heights": 18,
        "widths": [
          "66%",
          "10%",
          "10%",
          "14%",
        ],
      },
    },
  ],
]
`;

exports[`createPollPdfContentTable > should generate table for "byName" without groups 1`] = `
[
  {
    "table": {
      "body": [
        [
          {
            "color": "black",
            "fillColor": "white",
            "style": {
              "alignment": "left",
              "bold": true,
              "fontSize": 13,
              "margin": [
                0,
                5,
                0,
                0,
              ],
            },
            "text": "",
          },
          {
            "color": "black",
            "fillColor": "white",
            "style": "tableHeader",
            "text": "Yes
1",
            "type": "none",
          },
          {
            "color": "black",
            "fillColor": "white",
            "style": "tableHeader",
            "text": "No
1",
            "type": "none",
          },
          {
            "color": "black",
            "fillColor": "white",
            "style": "tableHeader",
            "text": "Invalid
1",
            "type": "none",
          },
        ],
        [
          {
            "style": {
              "alignment": "left",
            },
            "text": "Name of @user-1",
          },
          {
            "style": "tableBody",
            "text": "X",
          },
          {
            "style": "tableBody",
            "text": "",
          },
          {
            "style": "tableBody",
            "text": "",
          },
        ],
        [
          {
            "style": {
              "alignment": "left",
            },
            "text": "Name of @user-2",
          },
          {
            "style": "tableBody",
            "text": "",
          },
          {
            "style": "tableBody",
            "text": "",
          },
          {
            "style": "tableBody",
            "text": "X",
          },
        ],
        [
          {
            "style": {
              "alignment": "left",
            },
            "text": "Name of @user-3",
          },
          {
            "style": "tableBody",
            "text": "",
          },
          {
            "style": "tableBody",
            "text": "X",
          },
          {
            "style": "tableBody",
            "text": "",
          },
        ],
      ],
      "headerRows": 1,
      "heights": 18,
      "widths": [
        "66%",
        "10%",
        "10%",
        "14%",
      ],
    },
  },
]
`;

exports[`createPollPdfContentTable > should generate table for "open" with groups 1`] = `
[
  [
    [
      {
        "table": {
          "body": [
            [
              {
                "columns": [
                  {
                    "canvas": [
                      {
                        "color": "#ff0000",
                        "r1": 15,
                        "r2": 15,
                        "type": "ellipse",
                        "x": 15,
                        "y": 15,
                      },
                    ],
                    "width": 33,
                  },
                  {
                    "relativePosition": {
                      "y": 7.777777777777778,
                    },
                    "style": {
                      "alignment": "left",
                      "bold": true,
                      "fontSize": 13,
                    },
                    "text": "Group 1",
                  },
                ],
              },
              {
                "style": "tableHeader",
                "text": "Yes
1",
                "type": "none",
              },
              {
                "style": "tableHeader",
                "text": "No
0",
                "type": "none",
              },
              {
                "style": "tableHeader",
                "text": "Invalid
1",
                "type": "none",
              },
            ],
          ],
          "headerRows": 1,
          "heights": 18,
          "widths": [
            "66%",
            "10%",
            "10%",
            "14%",
          ],
        },
      },
      {
        "table": {
          "body": [
            [
              {
                "columns": [
                  {
                    "canvas": [
                      {
                        "color": "#0000ff",
                        "r1": 15,
                        "r2": 15,
                        "type": "ellipse",
                        "x": 15,
                        "y": 15,
                      },
                    ],
                    "width": 33,
                  },
                  {
                    "relativePosition": {
                      "y": 7.777777777777778,
                    },
                    "style": {
                      "alignment": "left",
                      "bold": true,
                      "fontSize": 13,
                    },
                    "text": "Group 2",
                  },
                ],
              },
              {
                "style": "tableHeader",
                "text": "Yes
0",
                "type": "none",
              },
              {
                "style": "tableHeader",
                "text": "No
1",
                "type": "none",
              },
              {
                "style": "tableHeader",
                "text": "Invalid
0",
                "type": "none",
              },
            ],
          ],
          "headerRows": 1,
          "heights": 18,
          "widths": [
            "66%",
            "10%",
            "10%",
            "14%",
          ],
        },
      },
      {
        "table": {
          "body": [
            [
              {
                "columns": [
                  {
                    "canvas": [
                      {
                        "color": "#ffaaff",
                        "r1": 15,
                        "r2": 15,
                        "type": "ellipse",
                        "x": 15,
                        "y": 15,
                      },
                    ],
                    "width": 33,
                  },
                  {
                    "relativePosition": {
                      "y": 7.777777777777778,
                    },
                    "style": {
                      "alignment": "left",
                      "bold": true,
                      "fontSize": 13,
                    },
                    "text": "Group 3",
                  },
                ],
              },
              {
                "style": "tableHeader",
                "text": "Yes
0",
                "type": "none",
              },
              {
                "style": "tableHeader",
                "text": "No
0",
                "type": "none",
              },
              {
                "style": "tableHeader",
                "text": "Invalid
0",
                "type": "none",
              },
            ],
          ],
          "headerRows": 1,
          "heights": 18,
          "widths": [
            "66%",
            "10%",
            "10%",
            "14%",
          ],
        },
      },
    ],
  ],
]
`;
